File: ios/Processors/PhotoIDMatchProcessor.swift
//
//  PhotoIDMatchProcessor.swift
//  ReactNativeCapfaceSdk
//
//  Created by Nayara Dias, Bruno Fialho and Daniel Sansão on 04/09/23.
//  Copyright © 2023 Capitual. All rights reserved.
//

import UIKit
import Foundation
import FaceTecSDK

class PhotoIDMatchProcessor: NSObject, Processor, FaceTecFaceScanProcessorDelegate, FaceTecIDScanProcessorDelegate, URLSessionTaskDelegate {
    var success = false
    var faceScanWasSuccessful = false
    var latestExternalDatabaseRefID: String!
    var data: NSDictionary!
    var latestNetworkRequest: URLSessionTask!
    var fromViewController: FaceTecViewController!
    var faceScanResultCallback: FaceTecFaceScanResultCallback!
    var idScanResultCallback: FaceTecIDScanResultCallback!
    private let principalKey = "photoIdMatchMessage";
    private let FaceThemeUtils: ThemeUtils! = ThemeUtils();
    
    
    init(sessionToken: String, fromViewController: FaceTecViewController, data: NSDictionary) {
        self.fromViewController = fromViewController
        self.latestExternalDatabaseRefID = self.fromViewController.getLatestExternalDatabaseRefID()
        self.data = data
        super.init()
        
        FaceTecCustomization.setIDScanUploadMessageOverrides(
            frontSideUploadStarted: self.FaceThemeUtils.handleMessage(self.principalKey, child: "frontSideUploadStarted", defaultMessage: "Uploading\nEncrypted\nID Scan"),                // Upload of ID front-side has started.
            frontSideStillUploading: self.FaceThemeUtils.handleMessage(self.principalKey, child: "frontSideStillUploading", defaultMessage: "Still Uploading...\nSlow Connection"),         // Upload of ID front-side is still uploading to Server after an extended period of time.
            frontSideUploadCompleteAwaitingResponse: self.FaceThemeUtils.handleMessage(self.principalKey, child: "frontSideUploadCompleteAwaitingResponse", defaultMessage: "Upload Complete"),             // Upload of ID front-side to the Server is complete.
            frontSideUploadCompleteAwaitingProcessing: self.FaceThemeUtils.handleMessage(self.principalKey, child: "frontSideUploadCompleteAwaitingProcessing", defaultMessage: "Processing\nID Scan"),       // Upload of ID front-side is complete and we are waiting for the Server to finish processing and respond.
            backSideUploadStarted: self.FaceThemeUtils.handleMessage(self.principalKey, child: "backSideUploadStarted", defaultMessage: "Uploading\nEncrypted\nBack of ID"),              // Upload of ID back-side has started.
            backSideStillUploading: self.FaceThemeUtils.handleMessage(self.principalKey, child: "backSideStillUploading", defaultMessage: "Still Uploading...\nSlow Connection"),          // Upload of ID back-side is still uploading to Server after an extended period of time.
            backSideUploadCompleteAwaitingResponse: self.FaceThemeUtils.handleMessage(self.principalKey, child: "backSideUploadCompleteAwaitingResponse", defaultMessage: "Upload Complete"),              // Upload of ID back-side to Server is complete.
            backSideUploadCompleteAwaitingProcessing: self.FaceThemeUtils.handleMessage(self.principalKey, child: "backSideUploadCompleteAwaitingProcessing", defaultMessage: "Processing\nBack of ID"),     // Upload of ID back-side is complete and we are waiting for the Server to finish processing and respond.
            userConfirmedInfoUploadStarted: self.FaceThemeUtils.handleMessage(self.principalKey, child: "userConfirmedInfoUploadStarted", defaultMessage: "Uploading\nYour Confirmed Info"),       // Upload of User Confirmed Info has started.
            userConfirmedInfoStillUploading: self.FaceThemeUtils.handleMessage(self.principalKey, child: "userConfirmedInfoStillUploading", defaultMessage: "Still Uploading...\nSlow Connection"), // Upload of User Confirmed Info is still uploading to Server after an extended period of time.
            userConfirmedInfoUploadCompleteAwaitingResponse: self.FaceThemeUtils.handleMessage(self.principalKey, child: "userConfirmedInfoUploadCompleteAwaitingResponse", defaultMessage: "Upload Complete"),     // Upload of User Confirmed Info to the Server is complete.
            userConfirmedInfoUploadCompleteAwaitingProcessing: self.FaceThemeUtils.handleMessage(self.principalKey, child: "userConfirmedInfoUploadCompleteAwaitingProcessing", defaultMessage: "Processing"),        // Upload of User Confirmed Info is complete and we are waiting for the Server to finish processing and respond.
            nfcUploadStarted: self.FaceThemeUtils.handleMessage(self.principalKey, child: "nfcUploadStarted", defaultMessage: "Uploading Encrypted\nNFC Details"),                   // Upload of NFC Details has started.
            nfcStillUploading: self.FaceThemeUtils.handleMessage(self.principalKey, child: "nfcStillUploading", defaultMessage: "Still Uploading...\nSlow Connection"),               // Upload of NFC Details is still uploading to Server after an extended period of time.
            nfcUploadCompleteAwaitingResponse: self.FaceThemeUtils.handleMessage(self.principalKey, child: "nfcUploadCompleteAwaitingResponse", defaultMessage: "Upload Complete"),                   // Upload of NFC Details to the Server is complete.
            nfcUploadCompleteAwaitingProcessing: self.FaceThemeUtils.handleMessage(self.principalKey, child: "nfcUploadCompleteAwaitingProcessing", defaultMessage: "Processing\nNFC Details"),         // Upload of NFC Details is complete and we are waiting for the Server to finish processing and respond.
            skippedNFCUploadStarted: self.FaceThemeUtils.handleMessage(self.principalKey, child: "skippedNFCUploadStarted", defaultMessage: "Uploading Encrypted\nID Details"),             // Upload of ID Details has started.
            skippedNFCStillUploading: self.FaceThemeUtils.handleMessage(self.principalKey, child: "skippedNFCStillUploading", defaultMessage: "Still Uploading...\nSlow Connection"),        // Upload of ID Details is still uploading to Server after an extended period of time.
            skippedNFCUploadCompleteAwaitingResponse: self.FaceThemeUtils.handleMessage(self.principalKey, child: "skippedNFCUploadCompleteAwaitingResponse", defaultMessage: "Upload Complete"),            // Upload of ID Details to the Server is complete.
            skippedNFCUploadCompleteAwaitingProcessing: self.FaceThemeUtils.handleMessage(self.principalKey, child: "skippedNFCUploadCompleteAwaitingProcessing", defaultMessage: "Processing\nID Details")    // Upload of ID Details is complete and we are waiting for the Server to finish processing and respond.
        );
        
        ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: true);
        
        let idScanViewController = FaceTec.sdk.createSessionVC(faceScanProcessorDelegate: self, idScanProcessorDelegate: self, sessionToken: sessionToken)
        
        FaceTecUtilities.getTopMostViewController()?.present(idScanViewController, animated: true, completion: nil)
    }
    
    func processSessionWhileFaceTecSDKWaits(sessionResult: FaceTecSessionResult, faceScanResultCallback: FaceTecFaceScanResultCallback) {
        fromViewController.setLatestSessionResult(sessionResult: sessionResult)
        
        self.faceScanResultCallback = faceScanResultCallback
        
        if sessionResult.status != FaceTecSessionStatus.sessionCompletedSuccessfully {
            if latestNetworkRequest != nil {
                latestNetworkRequest.cancel()
            }
            
            ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
            faceScanResultCallback.onFaceScanResultCancel()
            return
        }
        
        var parameters: [String : Any] = [:]
        if (self.data != nil) {
            parameters["data"] = self.data
        }
        parameters["faceScan"] = sessionResult.faceScanBase64
        parameters["auditTrailImage"] = sessionResult.auditTrailCompressedBase64![0]
        parameters["lowQualityAuditTrailImage"] = sessionResult.lowQualityAuditTrailCompressedBase64![0]
        parameters["externalDatabaseRefID"] = self.latestExternalDatabaseRefID
        
        var request = Config.makeRequest(url: "/enrollment-3d", httpMethod: "POST")
        request.httpBody = try! JSONSerialization.data(withJSONObject: parameters, options: JSONSerialization.WritingOptions(rawValue: 0))
        
        let session = URLSession(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: OperationQueue.main)
        latestNetworkRequest = session.dataTask(with: request as URLRequest, completionHandler: { data, response, error in
            
            guard let data = data else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return
            }
            
            guard let responseJSON = try? JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions.allowFragments) as? [String: AnyObject] else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return
            }
            
            guard let scanResultBlob = responseJSON["scanResultBlob"] as? String,
                  let wasProcessed = responseJSON["wasProcessed"] as? Bool else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return;
            }
            
            if wasProcessed == true {
                let message = self.FaceThemeUtils.handleMessage(self.principalKey, child: "successMessage", defaultMessage: "Liveness\nConfirmed");
                FaceTecCustomization.setOverrideResultScreenSuccessMessage(message);
                
                self.faceScanWasSuccessful = faceScanResultCallback.onFaceScanGoToNextStep(scanResultBlob: scanResultBlob);
            } else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                faceScanResultCallback.onFaceScanResultCancel()
                return;
            }
        })
        
        latestNetworkRequest.resume()
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 6) {
            if self.latestNetworkRequest.state == .completed { return }
            
            let message = self.FaceThemeUtils.handleMessage(self.principalKey, child: "uploadMessageIos", defaultMessage: "Still Uploading...");
            let uploadMessage:NSMutableAttributedString = NSMutableAttributedString.init(string: message);
            faceScanResultCallback.onFaceScanUploadMessageOverride(uploadMessageOverride: uploadMessage);
        }
    }
    
    func processIDScanWhileFaceTecSDKWaits(idScanResult: FaceTecIDScanResult, idScanResultCallback: FaceTecIDScanResultCallback) {
        fromViewController.setLatestIDScanResult(idScanResult: idScanResult)
        
        self.idScanResultCallback = idScanResultCallback
        
        if idScanResult.status != FaceTecIDScanStatus.success {
            if latestNetworkRequest != nil {
                latestNetworkRequest.cancel()
            }
            
            ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
            idScanResultCallback.onIDScanResultCancel()
            return
        }
        
        let minMatchLevel = 3
        
        var parameters: [String : Any] = [:]
        parameters["idScan"] = idScanResult.idScanBase64
        if idScanResult.frontImagesCompressedBase64?.isEmpty == false {
            parameters["idScanFrontImage"] = idScanResult.frontImagesCompressedBase64![0]
        }
        if idScanResult.backImagesCompressedBase64?.isEmpty == false {
            parameters["idScanBackImage"] = idScanResult.backImagesCompressedBase64![0]
        }
        parameters["minMatchLevel"] = minMatchLevel
        parameters["externalDatabaseRefID"] = self.latestExternalDatabaseRefID
        
        var request = Config.makeRequest(url: "/match-3d-2d-idscan", httpMethod: "POST")
        request.httpBody = try! JSONSerialization.data(withJSONObject: parameters, options: JSONSerialization.WritingOptions(rawValue: 0))
        
        let session = URLSession(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: OperationQueue.main)
        latestNetworkRequest = session.dataTask(with: request as URLRequest, completionHandler: { data, response, error in
            
            guard let data = data else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                idScanResultCallback.onIDScanResultCancel()
                return
            }
            
            guard let responseJSON = try? JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions.allowFragments) as? [String: AnyObject] else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                idScanResultCallback.onIDScanResultCancel()
                return
            }
            
            guard let scanResultBlob = responseJSON["scanResultBlob"] as? String,
                  let wasProcessed = responseJSON["wasProcessed"] as? Bool else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                idScanResultCallback.onIDScanResultCancel()
                return
            }
            
            if wasProcessed == true {
                FaceTecCustomization.setIDScanResultScreenMessageOverrides(
                    successFrontSide: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successFrontSide", defaultMessage: "ID Scan Complete"),                           // Successful scan of ID front-side (ID Types with no back-side).
                    successFrontSideBackNext: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successFrontSideBackNext", defaultMessage: "Front of ID\nScanned"),               // Successful scan of ID front-side (ID Types that do have a back-side).
                    successFrontSideNFCNext: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successFrontSideNFCNext", defaultMessage: "Front of ID\nScanned"),                // Successful scan of ID front-side (ID Types that do have NFC but do not have a back-side).
                    successBackSide: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successBackSide", defaultMessage: "ID Scan Complete"),                            // Successful scan of the ID back-side (ID Types that do not have NFC).
                    successBackSideNFCNext: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successBackSideNFCNext", defaultMessage: "Back of ID\nScanned"),                  // Successful scan of the ID back-side (ID Types that do have NFC).
                    successPassport: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successPassport", defaultMessage: "Passport Scan Complete"),                      // Successful scan of a Passport that does not have NFC.
                    successPassportNFCNext: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successPassportNFCNext", defaultMessage: "Passport Scanned"),                     // Successful scan of a Passport that does have NFC.
                    successUserConfirmation: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successUserConfirmation", defaultMessage: "Photo ID Scan\nComplete"),             // Successful upload of final IDScan containing User-Confirmed ID Text.
                    successNFC: self.FaceThemeUtils.handleMessage(self.principalKey, child: "successNFC", defaultMessage: "ID Scan Complete"),                                 // Successful upload of the scanned NFC chip information.
                    retryFaceDidNotMatch: self.FaceThemeUtils.handleMessage(self.principalKey, child: "retryFaceDidNotMatch", defaultMessage: "Face Didn’t Match\nHighly Enough"),       // Case where a Retry is needed because the Face on the Photo ID did not Match the User's Face highly enough.
                    retryIDNotFullyVisible: self.FaceThemeUtils.handleMessage(self.principalKey, child: "retryIDNotFullyVisible", defaultMessage: "ID Document\nNot Fully Visible"),       // Case where a Retry is needed because a Full ID was not detected with high enough confidence.
                    retryOCRResultsNotGoodEnough: self.FaceThemeUtils.handleMessage(self.principalKey, child: "retryOCRResultsNotGoodEnough", defaultMessage: "ID Text Not Legible"),            // Case where a Retry is needed because the OCR did not produce good enough results and the User should Retry with a better capture.
                    retryIDTypeNotSupported: self.FaceThemeUtils.handleMessage(self.principalKey, child: "retryIDTypeNotSupported", defaultMessage: "ID Type Mismatch\nPlease Try Again"),  // Case where there is likely no OCR Template installed for the document the User is attempting to scan.
                    skipOrErrorNFC: self.FaceThemeUtils.handleMessage(self.principalKey, child: "skipOrErrorNFC", defaultMessage: "ID Details\nUploaded")                          // Case where NFC Scan was skipped due to the user's interaction or an unexpected error.
                )
                
                self.success = idScanResultCallback.onIDScanResultProceedToNextStep(scanResultBlob: scanResultBlob)
            } else {
                ReactNativeCapfaceSdk.emitter.sendEvent(withName: "onCloseModal", body: false);
                idScanResultCallback.onIDScanResultCancel()
            }
        })
        
        latestNetworkRequest.resume()
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
        let uploadProgress: Float = Float(totalBytesSent) / Float(totalBytesExpectedToSend)
        if idScanResultCallback != nil {
            idScanResultCallback.onIDScanUploadProgress(uploadedPercent: uploadProgress)
        } else {
            faceScanResultCallback.onFaceScanUploadProgress(uploadedPercent: uploadProgress)
        }
    }
    
    func onFaceTecSDKCompletelyDone() {
        self.fromViewController.onComplete()
    }
    
    func isSuccess() -> Bool {
        return success
    }
}

File: android/src/main/java/com/capitual/processors/PhotoIDMatchProcessor.java
package com.capitual.processors;

import android.content.Context;

import androidx.annotation.NonNull;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.RequestBody;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;

import com.capitual.processors.helpers.ThemeUtils;
import com.capitual.reactnativecapfacesdk.ReactNativeCapfaceSdkModule;
import com.facebook.react.bridge.ReadableMap;
import com.facetec.sdk.*;

public class PhotoIDMatchProcessor extends Processor implements FaceTecFaceScanProcessor, FaceTecIDScanProcessor {
  private final String key = "photoIdMatchMessage";
  private final String exceptionJsonMessage = "Exception raised while attempting to parse JSON result.";
  private final String exceptionHttpMessage = "Exception raised while attempting HTTPS call.";
  private final String latestExternalDatabaseRefID;
  private final ReadableMap data;
  private final ReactNativeCapfaceSdkModule capFaceModule;
  private final ThemeUtils capThemeUtils = new ThemeUtils();
  private boolean success = false;
  private boolean faceScanWasSuccessful = false;

  public PhotoIDMatchProcessor(String sessionToken, Context context, ReactNativeCapfaceSdkModule capFaceModule,
      ReadableMap data) {
    this.capFaceModule = capFaceModule;
    this.latestExternalDatabaseRefID = this.capFaceModule.getLatestExternalDatabaseRefID();
    this.data = data;

    FaceTecCustomization.setIDScanUploadMessageOverrides(
        // Upload of ID front-side has started.
        capThemeUtils.handleMessage(key, "frontSideUploadStarted", "Uploading\nEncrypted\nID Scan"),
        // Upload of ID front-side is still uploading to Server after an extended period
        // of time.
        capThemeUtils.handleMessage(key, "frontSideStillUploading",
            "Still Uploading...\nSlow Connection"),
        // Upload of ID front-side to the Server is complete.
        capThemeUtils.handleMessage(key, "frontSideUploadCompleteAwaitingResponse",
            "Upload Complete"),
        // Upload of ID front-side is complete and we are waiting for the Server to
        // finish processing and respond.
        capThemeUtils.handleMessage(key, "frontSideUploadCompleteAwaitingProcessing",
            "Processing ID Scan"),
        // Upload of ID back-side has started.
        capThemeUtils.handleMessage(key, "backSideUploadStarted",
            "Uploading\nEncrypted\nBack of ID"),
        // Upload of ID back-side is still uploading to Server after an extended period
        // of time.
        capThemeUtils.handleMessage(key, "backSideStillUploading",
            "Still Uploading...\nSlow Connection"),
        // Upload of ID back-side to Server is complete.
        capThemeUtils.handleMessage(key, "backSideUploadCompleteAwaitingResponse",
            "Upload Complete"),
        // Upload of ID back-side is complete and we are waiting for the Server to
        // finish processing and respond.
        capThemeUtils.handleMessage(key, "backSideUploadCompleteAwaitingProcessing",
            "Processing Back of ID"),
        // Upload of User Confirmed Info has started.
        capThemeUtils.handleMessage(key, "userConfirmedInfoUploadStarted",
            "Uploading\nYour Confirmed Info"),
        // Upload of User Confirmed Info is still uploading to Server after an extended
        // period of time.
        capThemeUtils.handleMessage(key, "userConfirmedInfoStillUploading",
            "Still Uploading...\nSlow Connection"),
        // Upload of User Confirmed Info to the Server is complete.
        capThemeUtils.handleMessage(key, "userConfirmedInfoUploadCompleteAwaitingResponse",
            "Upload Complete"),
        // Upload of User Confirmed Info is complete and we are waiting for the Server
        // to finish processing and respond.
        capThemeUtils.handleMessage(key, "userConfirmedInfoUploadCompleteAwaitingProcessing",
            "Processing"),
        // Upload of NFC Details has started.
        capThemeUtils.handleMessage(key, "nfcUploadStarted",
            "Uploading Encrypted\nNFC Details"),
        // Upload of NFC Details is still uploading to Server after an extended period
        // of time.
        capThemeUtils.handleMessage(key, "nfcStillUploading",
            "Still Uploading...\nSlow Connection"),
        // Upload of NFC Details to the Server is complete.
        capThemeUtils.handleMessage(key, "nfcUploadCompleteAwaitingResponse",
            "Upload Complete"),
        // Upload of NFC Details is complete and we are waiting for the Server to finish
        // processing and respond.
        capThemeUtils.handleMessage(key, "nfcUploadCompleteAwaitingProcessing",
            "Processing\nNFC Details"),
        // Upload of ID Details has started.
        capThemeUtils.handleMessage(key, "skippedNFCUploadStarted",
            "Uploading Encrypted\nID Details"),
        // Upload of ID Details is still uploading to Server after an extended period of
        // time.
        capThemeUtils.handleMessage(key, "skippedNFCStillUploading",
            "Still Uploading...\nSlow Connection"),
        // Upload of ID Details to the Server is complete.
        capThemeUtils.handleMessage(key, "skippedNFCUploadCompleteAwaitingResponse",
            "Upload Complete"),
        // Upload of ID Details is complete and we are waiting for the Server to finish
        // processing and respond.
        capThemeUtils.handleMessage(key, "skippedNFCUploadCompleteAwaitingProcessing",
            "Processing\nID Details"));

    capFaceModule.sendEvent("onCloseModal", true);
    FaceTecSessionActivity.createAndLaunchSession(context, PhotoIDMatchProcessor.this, PhotoIDMatchProcessor.this,
        sessionToken);
  }

  public void processSessionWhileFaceTecSDKWaits(final FaceTecSessionResult sessionResult,
      final FaceTecFaceScanResultCallback faceScanResultCallback) {
    capFaceModule.setLatestSessionResult(sessionResult);

    if (sessionResult.getStatus() != FaceTecSessionStatus.SESSION_COMPLETED_SUCCESSFULLY) {
      NetworkingHelpers.cancelPendingRequests();
      faceScanResultCallback.cancel();
      capFaceModule.sendEvent("onCloseModal", false);
      capFaceModule.processorPromise.reject("The session status has not been completed!", "CapFaceInvalidSession");
      return;
    }

    JSONObject parameters = new JSONObject();
    try {
      if (this.data != null) {
        parameters.put("data", new JSONObject(this.data.toHashMap()));
      }
      parameters.put("faceScan", sessionResult.getFaceScanBase64());
      parameters.put("auditTrailImage", sessionResult.getAuditTrailCompressedBase64()[0]);
      parameters.put("lowQualityAuditTrailImage", sessionResult.getLowQualityAuditTrailCompressedBase64()[0]);
      parameters.put("externalDatabaseRefID", this.latestExternalDatabaseRefID);
    } catch (JSONException e) {
      e.printStackTrace();
      capFaceModule.sendEvent("onCloseModal", false);
      capFaceModule.processorPromise.reject("Exception raised while attempting to create JSON payload for upload.",
          "JSONError");
    }

    okhttp3.Request request = new okhttp3.Request.Builder()
        .url(Config.BaseURL + "/enrollment-3d")
        .headers(Config.getHeaders("POST"))
        .post(new ProgressRequestBody(
            RequestBody.create(MediaType.parse("application/json; charset=utf-8"), parameters.toString()),
            new ProgressRequestBody.Listener() {
              @Override
              public void onUploadProgressChanged(long bytesWritten, long totalBytes) {
                final float uploadProgressPercent = ((float) bytesWritten) / ((float) totalBytes);
                faceScanResultCallback.uploadProgress(uploadProgressPercent);
              }
            }))
        .build();

    NetworkingHelpers.getApiClient().newCall(request).enqueue(new Callback() {
      @Override
      public void onResponse(@NonNull Call call, @NonNull okhttp3.Response response) throws IOException {
        String responseString = response.body().string();
        response.body().close();
        try {
          JSONObject responseJSON = new JSONObject(responseString);
          boolean wasProcessed = responseJSON.getBoolean("wasProcessed");
          String scanResultBlob = responseJSON.getString("scanResultBlob");

          if (wasProcessed) {
            FaceTecCustomization.overrideResultScreenSuccessMessage = capThemeUtils.handleMessage(
                key, "successMessage",
                "Liveness\nConfirmed");
            faceScanWasSuccessful = faceScanResultCallback.proceedToNextStep(scanResultBlob);
          } else {
            faceScanResultCallback.cancel();
            capFaceModule.sendEvent("onCloseModal", false);
            capFaceModule.processorPromise.reject("CapFace SDK wasn't have to liveness values processed!",
                "CapFaceLivenessWasntProcessed");
          }
        } catch (JSONException e) {
          e.printStackTrace();
          faceScanResultCallback.cancel();
          capFaceModule.sendEvent("onCloseModal", false);
          capFaceModule.processorPromise.reject(exceptionJsonMessage, "JSONError");
        }
      }

      @Override
      public void onFailure(@NonNull Call call, @NonNull IOException e) {
        faceScanResultCallback.cancel();
        capFaceModule.sendEvent("onCloseModal", false);
        capFaceModule.processorPromise.reject(exceptionHttpMessage, "HTTPSError");
      }
    });
  }

  public void processIDScanWhileFaceTecSDKWaits(final FaceTecIDScanResult idScanResult,
      final FaceTecIDScanResultCallback idScanResultCallback) {
    capFaceModule.setLatestIDScanResult(idScanResult);

    if (idScanResult.getStatus() != FaceTecIDScanStatus.SUCCESS) {
      NetworkingHelpers.cancelPendingRequests();
      idScanResultCallback.cancel();
      capFaceModule.sendEvent("onCloseModal", false);
      capFaceModule.processorPromise.reject("The scan status has not been completed!", "CapFaceInvalidSession");
      return;
    }

    final int minMatchLevel = 3;

    JSONObject parameters = new JSONObject();
    try {
      parameters.put("externalDatabaseRefID", this.latestExternalDatabaseRefID);
      parameters.put("idScan", idScanResult.getIDScanBase64());
      parameters.put("minMatchLevel", minMatchLevel);

      ArrayList<String> frontImagesCompressedBase64 = idScanResult.getFrontImagesCompressedBase64();
      ArrayList<String> backImagesCompressedBase64 = idScanResult.getBackImagesCompressedBase64();
      if (frontImagesCompressedBase64.size() > 0) {
        parameters.put("idScanFrontImage", frontImagesCompressedBase64.get(0));
      }
      if (backImagesCompressedBase64.size() > 0) {
        parameters.put("idScanBackImage", backImagesCompressedBase64.get(0));
      }
    } catch (JSONException e) {
      e.printStackTrace();
      capFaceModule.sendEvent("onCloseModal", false);
      capFaceModule.processorPromise.reject(this.exceptionJsonMessage, "JSONError");
    }

    okhttp3.Request request = new okhttp3.Request.Builder()
        .url(Config.BaseURL + "/match-3d-2d-idscan")
        .headers(Config.getHeaders("POST"))
        .post(new ProgressRequestBody(
            RequestBody.create(MediaType.parse("application/json; charset=utf-8"), parameters.toString()),
            new ProgressRequestBody.Listener() {
              @Override
              public void onUploadProgressChanged(long bytesWritten, long totalBytes) {
                final float uploadProgressPercent = ((float) bytesWritten) / ((float) totalBytes);
                idScanResultCallback.uploadProgress(uploadProgressPercent);
              }
            }))
        .build();

    NetworkingHelpers.getApiClient().newCall(request).enqueue(new Callback() {
      @Override
      public void onResponse(@NonNull Call call, @NonNull okhttp3.Response response) throws IOException {
        String responseString = response.body().string();
        response.body().close();
        try {
          JSONObject responseJSON = new JSONObject(responseString);
          boolean wasProcessed = responseJSON.getBoolean("wasProcessed");
          String scanResultBlob = responseJSON.getString("scanResultBlob");

          if (wasProcessed) {
            FaceTecCustomization.setIDScanResultScreenMessageOverrides(
                // Successful scan of ID front-side (ID Types with no back-side).
                capThemeUtils.handleMessage(key, "successFrontSide",
                    "ID Scan Complete"),
                // Successful scan of ID front-side (ID Types that have a back-side).
                capThemeUtils.handleMessage(key, "successFrontSideBackNext",
                    "Front of ID\nScanned"),
                // Successful scan of ID front-side (ID Types that do have NFC but do not have a
                // back-side).
                capThemeUtils.handleMessage(key, "successFrontSideNFCNext",
                    "Front of ID\nScanned"),
                // Successful scan of the ID back-side (ID Types that do not have NFC).
                capThemeUtils.handleMessage(key, "successBackSide",
                    "ID Scan Complete"),
                // Successful scan of the ID back-side (ID Types that do have NFC).
                capThemeUtils.handleMessage(key, "successBackSideNFCNext",
                    "Back of ID\nScanned"),
                // Successful scan of a Passport that does not have NFC.
                capThemeUtils.handleMessage(key, "successPassport",
                    "Passport Scan Complete"),
                // Successful scan of a Passport that does have NFC.
                capThemeUtils.handleMessage(key, "successPassportNFCNext",
                    "Passport Scanned"),
                // Successful upload of final IDScan containing User-Confirmed ID Text.
                capThemeUtils.handleMessage(key, "successUserConfirmation",
                    "Photo ID Scan\nComplete"),
                // Successful upload of the scanned NFC chip information.
                capThemeUtils.handleMessage(key, "successNFC",
                    "ID Scan Complete"),
                // Case where a Retry is needed because the Face on the Photo ID did not Match
                // the User's Face highly enough.
                capThemeUtils.handleMessage(key, "retryFaceDidNotMatch",
                    "Face Didn't Match\nHighly Enough"),
                // Case where a Retry is needed because a Full ID was not detected with high
                // enough confidence.
                capThemeUtils.handleMessage(key, "retryIDNotFullyVisible",
                    "ID Document\nNot Fully Visible"),
                // Case where a Retry is needed because the OCR did not produce good enough
                // results and the User should Retry with a better capture.
                capThemeUtils.handleMessage(key, "retryOCRResultsNotGoodEnough",
                    "ID Text Not Legible"),
                // Case where there is likely no OCR Template installed for the document the
                // User is attempting to scan.
                capThemeUtils.handleMessage(key, "retryIDTypeNotSupported",
                    "ID Type Mismatch\nPlease Try Again"),
                // Case where NFC Scan was skipped due to the user's interaction or an
                // unexpected error.
                capThemeUtils.handleMessage(key, "skipOrErrorNFC",
                    "ID Details\nUploaded"));

            success = idScanResultCallback.proceedToNextStep(scanResultBlob);
            if (success) {
              capFaceModule.sendEvent("onCloseModal", false);
              capFaceModule.processorPromise.resolve(true);
            }
          } else {
            idScanResultCallback.cancel();
            capFaceModule.sendEvent("onCloseModal", false);
            capFaceModule.processorPromise.reject("CapFace SDK wasn't have to scan values processed!",
                "CapFaceScanWasntProcessed");
          }
        } catch (JSONException e) {
          e.printStackTrace();
          idScanResultCallback.cancel();
          capFaceModule.sendEvent("onCloseModal", false);
          capFaceModule.processorPromise.reject(exceptionJsonMessage, "JSONError");
        }
      }

      @Override
      public void onFailure(@NonNull Call call, @NonNull IOException e) {
        idScanResultCallback.cancel();
        capFaceModule.sendEvent("onCloseModal", false);
        capFaceModule.processorPromise.reject(exceptionHttpMessage, "HTTPSError");
      }
    });
  }

  public boolean isSuccess() {
    return this.success;
  }
}
